### PR TITLE
Added support hardswish function

### LINKED ISSUE #92: Operator support for `F.hardswish`
## 🚀 Feature

Implement HardSwish activation function.

### Motivation

Relatively easy activation function implementation as a good first issue as nikitaved suggested under #64 

### Pitch

Add HardSwish (x * ReLU6(x + 3) / 6) leveraging existing ReLU6 support.

cc @apaz-cli

### PR DESCRIPTION
**Pull request by shaharelys**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? Yes, #92 
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
Yes
btw, the url should be fixed to 'https://github.com/Lightning-AI/pytorch-lightning/blob/**master**/.github/CONTRIBUTING.md'.
- [ ] Did you make sure to update the docs? Not yet
- [x] Did you write any new necessary tests? Yes, only in `opinfo.py`

</details>

## What does this PR do?
Implements the HardSwish activation function.

Fixes #92 

## PR review

This PR is open for review, though I'm not entirely certain it's 100% complete, as this is my first contribution. All tests have passed, but `docs` has not been updated. Feedback from the community is welcomed!

## Did you have fun?

Yes! 🙃

### CODE REVIEW

**Comment by t-vi** [comment]
@shaharelys Thank you, so excited to see you with a PR here.

**Comment by shaharelys** [comment]
Hey guys! @nikitaved @t-vi, I would love to get your take on this whenever possible. In the meantime, I'll start working on the one_hot issue at #64.

**Comment by t-vi** [comment]
@shaharelys, Apologies for the delay. The PR seems great at first sight,  but I wanted to have @nikitaved take a more thorough look and we've hit a very long weekend with bank holidays both Friday and today.

**Comment by t-vi** [comment]
While you are waiting, could you look into how to only test for the floating point inputs (on cuda?):
```
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_nvfuser_cuda_int64 - RuntimeError: "hardswish_cuda" not implemented for 'Long'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_nvfuser_cuda_int32 - RuntimeError: "hardswish_cuda" not implemented for 'Int'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_torch_cuda_int64 - RuntimeError: "hardswish_cuda" not implemented for 'Long'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_torch_cuda_int16 - RuntimeError: "hardswish_cuda" not implemented for 'Short'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_torch_cuda_int32 - RuntimeError: "hardswish_cuda" not implemented for 'Int'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_torch_cuda_uint8 - RuntimeError: "hardswish_cuda" not implemented for 'Byte'
FAILED thunder/tests/test_ops.py::test_core_vs_torch_consistency_hardswish_torch_cuda_int8 - RuntimeError: "hardswish_cuda" not implemented for 'Char'
```

**Comment by t-vi** [comment]
I usually try to run this locally with `pytest-3 thunder/tests/test_ops.py -k hardswish` or so.

**Comment by shaharelys** [comment]
Okay, I believe this issue can be addressed by updating the following:

    ```
    # PyTorch does not support CPU integer types hardswish
        DecorateInfo(
            pytest.mark.xfail,
            "test_core_vs_torch_consistency",
            dtypes=(datatypes.int16, datatypes.int32, datatypes.int64, datatypes.int8, datatypes.uint8),
            devicetypes=(devices.DeviceType.CPU,),
        ),
        ```
to:
    ```
    # PyTorch does not support integer types for both the CPU and CUDA hardswish
        DecorateInfo(
            pytest.mark.xfail,
            "test_core_vs_torch_consistency",
            dtypes=(datatypes.int16, datatypes.int32, datatypes.int64, datatypes.int8, datatypes.uint8),
        ),
    ```
 
I've made this adjustment, and the issue should now be resolved.

Btw @t-vi, how could I have identified this error on my own? I ran `pytest thunder/tests/test_ops.py -k hardswish -v` with `pytest ==7.4.4` and only got `16 passed, 3 skipped, 2074 deselected, 6 xfailed, 1 xpassed, 1 warning`

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -1208,6 +1208,12 @@ def relu6(a: TensorProxy, /, inplace: bool = False) -> TensorLike:
     return clamp(a, 0, 6)
 
 
+@torchsymbol(torch.nn.functional.hardswish, id="torch.hardswish", is_method=False)
+def hardswish(a: TensorProxy, /, inplace: bool = False) -> TensorLike:
```
Comment (mruberry) [inline thunder/torch/__init__.py:1212]: This looks pretty good; following the above comment, please also add a check that `a` has a floating point dtype

File: thunder/tests/opinfos.py [hunk 2]
```diff
@@ -1598,6 +1598,52 @@ def relu6_error_generator(op, device, dtype=torch.float32, **kwargs):
...
+    torch_reference=_elementwise_unary_torch(torch.nn.functional.hardswish),
+    test_directives=(
+        # PyTorch does not support bool for both CPU and CUDA hardswish
+        DecorateInfo(
+            pytest.mark.xfail,
+            "test_core_vs_torch_consistency",
+            dtypes=(datatypes.bool8,),
+        ),
+        # PyTorch does not support complex types for both the CPU and CUDA hardswish
+        DecorateInfo(
```
[the remark this answers is not in the document]
Comment (mruberry) [reply thunder/tests/opinfos.py:1619]: (We could also let hardswish promote inexact types to floating point, of course, like many other operators do, but that's pretty unlikely to come up)

**Comment by mruberry** [comment]
Looks pretty good, @shaharelys! I made a recommendation to limit the operation's supported dtypes. I think the operation isn't defined on complex types, and while defined on exact types they're not very important to support.

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -1211,6 +1211,9 @@ def relu6(a: TensorProxy, /, inplace: bool = False) -> TensorLike:
 @torchsymbol(torch.nn.functional.hardswish, id="torch.hardswish", is_method=False)
 def hardswish(a: TensorProxy, /, inplace: bool = False) -> TensorLike:
     utils.check(not inplace, lambda: f"hardswish only supports inplace=False", exception_type=NotImplementedError)
+    utils.check(a.dtype in (torch.float16, torch.float32, torch.float64),
```
Comment (mruberry) [inline thunder/torch/__init__.py]: This is a good update, and the PR is almost ready to be merged.

Two final tweaks:

1) This won't detect all floating point dtypes, like bfloat16, or future dtypes. Instead, use this function:

https://github.com/Lightning-AI/lightning-thunder/blob/56bc01b8e17ba9551395a23c639fae79bc9c25b3/thunder/core/dtypes.py#L440

2) The error message looks good, but don't set the `exception_type` or set it to `ValueError`. It's an interesting idea to consider dtype mismatch as a `TypeError`, but currently we don't consider the datatype of a tensor as part of its "type" (which is just tensor)

**Review by mruberry — APPROVED** [review]
Thanks @shaharelys! Looks great. The CI failure is an unrelated flaky failure.

**Comment by shaharelys** [comment]
Super excited about this! @mruberry Thanks for the support! Will now move back to the one_hot at #64

**Pull request merged**